### ISSUE TITLE
Mangles certain file:// repository URLs

### ISSUE THREAD

**Issue by robx** [haskell/cabal#7937]
**Describe the bug**
If you configure a repo with a URL of the form `file://foo/bar` (as opposed to the intended `file:///foo/bar`), `foo` is silently stripped.

**To Reproduce**

```
$ cat cabal.config
repository test-local-repo
  url: file://dir/cabal-repo
$ cabal --config-file=cabal.config v2-update
Downloading the latest package list from test-local-repo
/cabal-repo/00-index.tar.gz: copyFile: does not exist (No such file or directory)
```

**Expected behavior**
I expect an error for the missing leading flag (or perhaps to look for the repo in `/dir/cabal-repo`)

**System information**
 - Operating system: macos
 - `cabal` 3.6.2.0

**Comment by Mikolaj** [comment]
Wow. So you intend to use an absolute path, but what you wrote is a relative path, but instead of using the relative path, cabal drops letters until it finds something that looks like an absolute path again? That's certainly a bug.

Are relative paths permitted in this context? Should they be? Are they useful in this context? Should we signal an error, use the relative path or turn relative to absolute by adding `/` at the start?

**Comment by robx** [comment]
Huh, it appears that `file://` URLs have a host in there, so `file://foo/bar` means the file `/bar` on the host `foo`: https://en.wikipedia.org/wiki/File_URI_scheme.

```
> let Just x = parseURI "file://hello/world"
> uriAuthority x
Just (URIAuth {uriUserInfo = "", uriRegName = "hello", uriPort = ""})
```

I guess the best thing to do might be to error out if the URI has a non-empty `uriRegName` that's not `localhost`, but not really sure what's best. It might be valid to judge this "not a bug"...

**Comment by Mikolaj** [comment]
And can something different that `localhost` make sense here? What protocol would be used to access the non-local host?

**Comment by robx** [comment]
I don't think it can make sense in the context of `cabal`. I imagine the idea might be some context in which e.g. two hosts `foo` and `bar` share a set of URIs, and e.g. `foo` has `bar`'s filesystem mounted at `/mount/bar`, so it knows it can translate `file://bar/etc/passwd` to the local path `/mount/bar/etc/passwd`... I doubt use of `file:` URIs with meaningful hostnames are widespread.

**Comment by Mikolaj** [comment]
In that case, I think it's useful to warn and perhaps even error out, suggesting the extra `/`.
